Fix urlize markdown extension

azaghal opened the pull request:
This PR revamps the way the urlize markdown extension is implemented. It is *not* backwards compatible, though. While testing the old implementation, there was a number of issues I ran into (like lots of inline samples not working).

One of the big changes is that the link must be a standalone word (e.g. either at beginning/end of text or surrounded by spaces). This should help avoid some issues.

In general, though, one could see this extension (even the old implementation) as somewhat fragile/too optimistic. For example, if you have an OID 1.2.3.4, it'll get recognised as a link.

On a positive note, the extension has a bit more extensive tests in place now.

azaghal commented on src/wiki/plugins/links/mdx/urlize.py:
```diff
@@ -2,98 +2,138 @@
+
+        matched_string = m.group(0)
 
-        if url.startswith('<'):
```
Actually, this one might not be correctly handled in my new code - what was the intention of this specific check in the original code? I might need to remove special handling for URLs like '<example.com>' or '(example.com)'? Not sure.

benjaoming replied:
I'm not sure if using `urlize` should cover such. After all, it's meant as a simple shortcut to make certain things clickable. It would be nice if it didn't fire on too many objects and created useless broken links without the user being able to disable it. For all other use cases that it doesn't cover, we can just refer people to the true Markdown syntax: `[text to click](http://whatever.horse)`

azaghal replied:
Ok, I'll update code to not mess with these characters at all.

azaghal replied:
I ended-up including the `()` and `<>` handling, but in a more robust way.

benjaoming commented:
Haven't arrived to reviewing this quite yet, sorry -- might be a couple of days.

> It is not backwards compatible

This part, could you clarify?

benjaoming commented:
Really exciting and needed improvement btw!

azaghal commented:
Apropos backwards incompatibilities, well, the following comes to my mind:

- Old code was not able to pass a number of the listed inline samples. It'd produce invalid results, so in a way my code would pass some of those.
- I've removed requirement for matching only `www.DOMAIN` as a FQDN link. Now any kind of FQDN will work.
- I've introduced requirement to have the link standing as standalone string. E.g. it must be at text beginning, text ending, line beginning, or have spaces (any sort of spaces) around it. This helps avoid some issues related to partial matches and the like, for example `-invalid.example.com` would end-up being `<p>-invalid<a href="http://example.com/">example.com</a><p>` in the old code, while with my code it will not end-up being matched at all (can't begin FQDN with a leading dash).
- The matching logic for IPv6 has been changed, since I ran into some kind of loop with the old pattern. Not sure if my new code is the right thing to do. Had a look at some IPv6 regexes out there, but I think they're major PITA to understand and maintain long-term.

I guess one way to check would be to run the new tests against the old code to see what the differences are.

benjaoming commented:
@azagha Thanks for explaining!l That's really cool, thanks! Am not sure how later versions of Django's `urlize` template filter works, but perhaps it is also worth a look.

Sounds like the backwards incompatibilities are desirable, so I wouldn't call them incompatibilities -- maybe just "improvements" :)

atombrella commented on tests/plugins/links/test_urlize.py:
```diff
@@ -0,0 +1,231 @@
+    def test_positive_matches(self, markdown_text, expected_output):
+        assert self.md.convert(markdown_text) == expected_output
+
+    @pytest.mark.parametrize("markdown_text, expected_output", FIXTURE_NEGATIVE_MATCHES)
```
Isn't this equivalent/comparable to `unittest.subTest`? I'm not sure how @benjaoming feels about different styles in the test suite, but I think choosing one consistent style is easier to maintain.

benjaoming replied:
Yup, I would agree that if `unittest` can do it, that's option 1.. then what `pytest` can offer would be second choice. But I think it's highly highly imaginative to think that one would want to change from `pytest` in a near future. Seems unlikely that a better test suite pops up or that `pytest` makes breaking API changes.

azaghal replied:
I'll have a look at what unittest has to offer in terms of parametrisation. This stuff could be done in loop as well, but thought it was a nice thing to take advantage of functionality.

azaghal replied:
Just had a look at it, that's a nice addition to unittest. I still find pytest a little bit superior, but better to keep everything consistent :)

azaghal replied:
Ok, I take it back a bit. There's a [known missing support for `subTest`](https://github.com/pytest-dev/pytest/issues/1367) in `pytest`.

So, with pytest variant all possible things will be run even if some of them fail, with `pytest + subTest`, tests will stop after first failure.

So... Pick your poison, gentlemen :)

I could "worst case" convert everything to TestCase etc, but still use `pytest`'s parametrisation.

benjaoming replied:
@azaghal then I think it's totally fair to leave it as it is :) I don't want to impose some really rigid way of cutting out obvious benefits of using `pytest` -- after all, it's there to improve testing.

azaghal replied:
Sheesh, ok, another update [from official pytest docs](https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses) - the parametrize thing is not supported.

So... Might need to stick to `pytest`-style unit test unless you're fine with test stopping on first failure?

Again a matter of choose your poison thing, just let me know :)

azaghal replied:
I'll be leaving the tests as `pytest`-tailored.

atombrella commented on tests/plugins/links/test_urlize.py:
```diff
@@ -0,0 +1,231 @@
+
+    @pytest.mark.parametrize("markdown_text, expected_output", FIXTURE_NEGATIVE_MATCHES)
+    def test_negative_matches(self, markdown_text, expected_output):
+        assert self.md.convert(markdown_text) == expected_output
```
See comment above, use `self.assertEqual` instead? This is what all the other tests use.

benjaoming replied:
Actually, when you are running `pytest`, it will catch these assertion errors and display the problems just as cleanly as `self.assertEqual`. Maybe even better. I don't want to say that it's necessarily better than the other, because it implies fixing up the remaining codebase or having some policy.. I would think both are fine :)

azaghal replied:
If nothing else in the test code uses this, I'll convert this into `unittest` test. Consistency is good. Or, as I like to put it - you want to introduce something new? Then go ahead and do it _everywhere_ :)

azaghal replied:
Went for ease of use instead of consistency. I'll trump over my own words and not do it everywhere, though...

benjaoming commented on src/wiki/plugins/links/mdx/urlize.py:
```diff
@@ -2,98 +2,138 @@
->>> md.convert('go to <http://example.com> now!')
-'<p>go to <a href="http://example.com">http://example.com</a> now!</p>'
+    # Leading protocol specification.
+    r'(?P<protocol>http://|https://|ftp://|ftps://|)'
```
Could also be `onion` :P

Actually this might as well be agnostic to the protocol? https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax

azaghal replied:
I'll incorporate this change as well.

benjaoming commented on src/wiki/plugins/links/mdx/urlize.py:
```diff
@@ -2,98 +2,138 @@
+    r'([A-F0-9]{1,4}:){1,6}:([A-F0-9]{1,4}){1,6}|'  # IPv6, zeros in middle removed.
+    r'\[?([A-F0-9]{1,4}:){1,6}:\]?|'  # IPv6, trailing zeros removed
+    r'\[?::\]?|'  # IPv6, just "empty" address
+    r'([A-Z0-9]([A-Z0-9-]{0,61}[A-Z0-9])?\.)+([A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # FQDN
```
Great! We retain our `.horse` TLD support :)

azaghal replied:
I actually had to go ahead and verify there _is_ a `.horse` TLD. And congratulations - I'm adding a test for it.

benjaoming commented on src/wiki/plugins/links/mdx/urlize.py:
```diff
@@ -2,98 +2,138 @@
 
-    """ Return a link Element given an autolink (`http://example/com`). """
+        # Ensure links are matched only if they stand on their own to avoid bad matches etc.
+        return re.compile(r'^(|.*?\s)%s(\s.*?|)$' % URLIZE_RE, re.DOTALL | re.UNICODE | re.IGNORECASE)
```
There's already the group called `end` in `URLIZE_RE`, perhaps somehow merge the remaining regular expression logic with that and add a group `start` ?

azaghal replied:
I think begin/end will end-up being dropped once I remove the `()` and `<>` syntax support.

teury commented:
![urlize](https://user-images.githubusercontent.com/5114048/48092961-723a0c00-e1e4-11e8-8113-d1a55122b1d4.jpg)

Amazing work.

I see a small error that creates links after a point.

benjaoming commented:
@teury thanks opened an issue for that - are you able to work on it? :)
